### DISCUSSION TITLE
ghc 9.4.3 support

### DISCUSSION THREAD

**Discussion by shinzui** [srid/haskell-flake#59] [Q&A]
I've been curious about `haskell-flake` and `flake-parts` for some time, so I tried to convert an existing project that's already using nix flakes to `haskell-flake`, but I wasn't able to get far, unfortunately. 

Here is the smallest broken flake file. 

```nix
{
  description = "project";
  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/nixpkgs-unstable";
    flake-parts.url = "github:hercules-ci/flake-parts";
    haskell-flake.github:srid/haskell-flake";
  };
  outputs = inputs@{ self, nixpkgs, flake-parts, haskell-flake, ... }:
    flake-parts.lib.mkFlake { inherit inputs; } {
      systems = nixpkgs.lib.systems.flakeExposed;
      imports = [
        haskell-flake.flakeModule
      ];
      perSystem = { config, pkgs, ... }: {
        haskellProjects.default = {
          haskellPackages = pkgs.haskell.packages.ghc943;
        };
      };
    };
}
```

I get the following error:

<img width="950" alt="image" src="https://user-images.githubusercontent.com/519/213964046-4008850f-a8f8-463a-b993-93b8888cb740.png">

Commenting out the `haskellPackages` line works, but I get ghc 9.2 instead of 9.4 which this project needs. 

Am I doing something wrong?

**Comment by srid** [comment]
I don't think you are doing anything wrong. nixpkgs provides 9.2 by default (ie., `pkgs.haskellPackages`), which should have most supported packages.  When you use a higher version, like 9.4.4, you would expect a bunch of packages to fail, like `ghcid` is failing here. Since it it is only tests, you can disable it as follows:

```nix
haskellPackages.default = {
  overrides = self: super: with pkgs.haskell.lib; {
    ghcid = dontCheck super.ghcid;
  };
}
```

You want to override other failing packages in a similar fashion. `callHackage` is another, if you want to use a specific version from Hackage that is known to build. Or you can provide a flake input of the git repo containing the latest code.

Alternatively, you can use @locallycompact's https://horizon-haskell.net/ overlay in haskell-flake. There was an attempt here, https://github.com/srid/haskell-template/pull/55/files - I don't know where that stands right now ...

**Reply by srid** [reply]
Here's an example of overrides (back when 9.0 was the default in nixpkgs):

https://github.com/srid/ema/blob/61faae56aa0f3c6ca815f344684cc566f6341662/flake.nix#L58-L61

**Reply by srid** [reply]
Oh, and as the linked PR shows - you can also set the default build tools in `buildTools` to the value `null` to disable them. i.e,

```nix
haskellPackages.default = {
  buildTools = hp: {
     haskell-language-server = null;
     hlint = null;
     ghcid = null;
  };
}
```

See also https://github.com/srid/haskell-flake/issues/37

**Reply by locallycompact** [reply]
horizon-platform now exports a full package set, so you would replace the package set with `horizon-platform.legacyPackages.${system}` rather than use the overrides field

**Reply by srid** [reply]
@locallycompact what if I wanted to use my own nixpkgs *and* use the Haskell 9.4 package set provided by Horizon?

**Reply by locallycompact** [reply]
I think if you try to merge two full package sets then you'd need to define custom functions to do that. I could techincally re-export the overlay expression again though, but it always felt a bit funny to me.

**Reply by shinzui** [reply]
Does the horizon platform support `aarch64-darwin`?

**Reply by locallycompact** [reply]
I'm unable to support darwin directly because I don't own any hardware, but it might be somewhat close to working. It does evaluate on darwin, but I'm told it doesn't build, as I've seen with the setup they have in this issue for flora: https://github.com/flora-pm/flora-server/issues/275

So it seems like profiling is not enabled for base on darwin in the configuration there. Either we need to find a way to turn profiling on for base on darwin, or map an override over the entire package set to disable profiling - that might build.

**Comment by shinzui** [comment]
Thank you for replying, @srid. What surprised me is that the project was already using nix for getting all my build tools 
(`pkgs.haskell.packages.ghc943.haskell-language-server`, `pkgs.haskell.compiler.ghc943`). I must familiarize myself with the `cabal2nix` infrastructure to understand what's happening. I'll also try the `horizon-platform` to see if it resolves my problem.
